File: codebasics/BST/binarySearchTree.js
class BinarySearchTree {

    constructor (data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }


    addChild (value) {
        if (value == this.data) {
            return;
        }

        if (value > this.data) {
            if (this.right) {
                this.right.addChild(value);
            } else {
                this.right = new BinarySearchTree(value);
            }
        } else {
            if (this.left) {
                this.left.addChild(value);
            } else {
                this.left = new BinarySearchTree(value);
            }
        }
    }


    searchNode (value) {
        if (value == this.data) {
            return true;
        }

        if (value > this.data) {
            if (this.right) {
                return this.right.searchNode(value);
            } else {
                return false;
            }
        } else {
            if (this.left) {
                return this.left.searchNode(value);
            } else {
                return false;
            }
        }
    }


    inOrderTraversal () {
        let treeSorted = [];

        if (this.left) {
            treeSorted = treeSorted.concat(this.left.inOrderTraversal());
        }

        treeSorted.push(this.data);

        if (this.right) {
            treeSorted = treeSorted.concat(this.right.inOrderTraversal());
        }

        return treeSorted;
    }

}


function buildTree (array) {
    console.log("Given array to convert into a tree ->", array);
    const root = new BinarySearchTree(array[0]);

    for (let i = 1; i < array.length; i++) {
        root.addChild(array[i]);
    }

    return root;
}


/* Example usage */
let randomArray = [];
for (let i = 0; i < 8; i++) {
    randomArray.push(Math.floor(Math.random() * 20 + 1));
}

const generalTree = buildTree(randomArray);
console.log("Tree sorted ->", generalTree.inOrderTraversal());

const elementToFound = 8
console.log(`Element ${8} was found in the tree? ->`, generalTree.searchNode(elementToFound));

File: codebasics/BST/binary_search_tree.py
import random
from typing import Any

class BinarySearchTreeNode:
    def __init__(self, data: Any) -> None:
        self.data = data
        self.left = None
        self.right = None


    def add_child(self, data: Any) -> None:
        if data == self.data:
            return
        
        if data < self.data:
            if self.left:
                self.left.add_child(data)
            else:
                self.left = BinarySearchTreeNode(data)
        else:
            if self.right:
                self.right.add_child(data)
            else:
                self.right = BinarySearchTreeNode(data)


    def in_order_traversal(self) -> list:
        array_sorted = list()

        if self.left:
            array_sorted += self.left.in_order_traversal()

        array_sorted.append(self.data)

        if self.right:
            array_sorted += self.right.in_order_traversal()

        return array_sorted
        

def build_tree(array: list) -> BinarySearchTreeNode:
    print("Array para convertir en arbol ->", array)
    root = BinarySearchTreeNode(array[0])

    for i in range(1, len(array)):
        root.add_child(array[i])

    return root
        

if __name__ == '__main__':
    random_array = [random.randint(1, 21) for _ in range(8)]

    create_tree = build_tree(random_array)
    print("El arbol binario ordenado es ->", create_tree.in_order_traversal())
